runtime/src/secondary_index.rs

```diff
@@ -99,12 +103,125 @@ impl SecondaryIndexEntry for RwLockSecondaryIndexEntry {
     }
 }
 
+#[derive(Debug, Default)]
+struct HierarchicalOrderedMap<L1, L2>
+where
+    L1: Default + Sync + Send + Copy + PartialEq + Ord,
+    L2: Default + Sync + Send + Copy + PartialEq + Ord,
+{
+    capacity: usize,
+    map: Vec<(L1, L2)>,
+    excluded_keys: Vec<L1>,
+}
+
+impl<L1, L2> HierarchicalOrderedMap<L1, L2>
+where
+    L1: Default + Sync + Send + Copy + PartialEq + Ord,
+    L2: Default + Sync + Send + Copy + PartialEq + Ord,
+{
+    pub fn new(capacity: usize, excluded_keys: &[L1]) -> Self {
+        Self {
+            capacity,
+            map: Vec::new(),
+            excluded_keys: Vec::from(excluded_keys),
+        }
+    }
+    fn get_map(&self) -> &Vec<(L1, L2)> {
+        &self.map
+    }
+    fn sort_slice_by_value(&mut self, slice_key: &L1) {
+        // Obtain a slice of mutable references to all elements with the same key
+        let mut sub_slice = self
+            .map
+            .split_mut(|(k, _)| *k != *slice_key)
+            .flatten()
+            .collect_vec();
```

Comment:
Why do these need itertools-specific collect_vec method?

Reply:
It seemed more convenient, why is this not supposed to be used?

Reply:
removed.

Reply:
I don't follow why it's more convenient than Iterator::collect().
Generally we ought to have good reasons to use a 3rd-party library over common, standard implementations. We do use Itertools in places, because it offers some functionality that std::iter::Iterator doesn't. But this doesn't seem to be the case here, unless I'm missing something.